tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t

```diff
@@ -0,0 +1,109 @@
+Setup
+
+  $ source "$TESTDIR"/_setup.sh
+
+Generate metadata file with 250 rows.
+
+  $ echo "strain	date	location" > metadata.tsv
+  $ for i in $(seq 1 50); do
+  >     echo "2000A_$i	2000	A" >> metadata.tsv
+  >     echo "2000B_$i	2000	B" >> metadata.tsv
+  >     echo "2001A_$i	2001	A" >> metadata.tsv
+  >     echo "2001B_$i	2001	B" >> metadata.tsv
+  >     echo "2002B_$i	2002	B" >> metadata.tsv
+  > done
+
+Weight locations A:B as 2:1. This is reflected in target_group_sizes.tsv below.
+
+  $ cat >weights.tsv <<~~
+  > location	weight
+  > A	2
+  > B	1
+  > ~~
+
+  $ ${AUGUR} filter \
+  >   --metadata metadata.tsv \
+  >   --group-by location \
+  >   --group-by-weights weights.tsv \
+  >   --subsample-max-sequences 100 \
+  >   --subsample-seed 0 \
+  >   --output-group-by-sizes target_group_sizes.tsv \
+  >   --output-strains strains.txt 2>/dev/null
+
+  $ cat target_group_sizes.tsv
+  location	weight	_augur_filter_target_size_int
+  A	2	67
+  B	1	33
+
+Using 1:1 weights is similarly straightforward, with 50 sequences from each location.
+
+  $ cat >weights.tsv <<~~
+  > location	weight
+  > A	1
+  > B	1
+  > ~~
+
+  $ ${AUGUR} filter \
+  >   --metadata metadata.tsv \
+  >   --group-by location \
+  >   --group-by-weights weights.tsv \
+  >   --subsample-max-sequences 100 \
+  >   --subsample-seed 0 \
+  >   --output-group-by-sizes target_group_sizes.tsv \
+  >   --output-strains strains.txt 2>/dev/null
+
+  $ cat target_group_sizes.tsv
+  location	weight	_augur_filter_target_size_int
+  A	1	50
+  B	1	50
+
+Keep the 1:1 location weighting, but add uniform sampling on year.
+The uniform sampling happens "within" each weighted column value, so the 1:1
+location weighting is reflected even though there is an imbalance in years
+available per location.
+
+  $ ${AUGUR} filter \
+  >   --metadata metadata.tsv \
+  >   --group-by year location \
+  >   --group-by-weights weights.tsv \
+  >   --subsample-max-sequences 100 \
+  >   --subsample-seed 0 \
+  >   --output-group-by-sizes target_group_sizes.tsv \
+  >   --output-strains strains.txt 2>/dev/null
+
+  $ cat target_group_sizes.tsv
+  year	location	weight	_augur_filter_target_size_int
+  2000	A	0.5	25
+  2000	B	0.3333333333333333	16
+  2001	A	0.5	25
+  2001	B	0.3333333333333333	16
+  2002	B	0.3333333333333333	17
+
+If a single sequence is added for group (2002,A), the weighting now appears
+"equal" among all years and locations.
+
+However, there is only 1 sequence available in (2002,A), much lower than the
+requested 17, so the total number of sequences outputted is lower than requested.
```

Comment:
This is an interesting example! It's similar to [our previous conversation](https://github.com/nextstrain/augur/pull/1454#discussion_r1596214631) about how to divide up samples within `location=A` via the year axis:

> Put another way, I think extra group-bys should be used to partition the data within the existing weighted groups, and thus the temporal distribution of data from a location will only affect the sampling of data from that location not other locations.

In this context we aim for uniform sampling within "A", so around 16 x 2000, 16 x 2001, 16 x 2002, and since there's only one 2002/A sample we undersample A and thus don't get our desired 1:1 A:B ratio. That's kinda weird. The alternate approach (at least, that I can think of) is to in-fill samples until we get to the target number for A. So we'd start with `{2000: 16, 2001: 16, 2002: 1}` which totals 33, but we want 50, so we'd add iteratively add samples with a ratio of (1/3 x 2000, 1/3 x 2001, 1/3 x 2002) until we reached 50 samples which would be circa `{2000: 25, 2001: 24, 2002: 1}`. 

No matter what approach we use, we should add output describing what's happening. I.e.

```diff
NOTE: Weights were not provided for the column 'year'.
Using equal weights across values in that column
+ NOTE: Within group "location=A" we uniformly sampled across "year"
+ however there were insufficient samples for "year=2002" resulting
+ in an undersampling of "location=A". 
```

```diff
NOTE: Weights were not provided for the column 'year'.
Using equal weights across values in that column
+ NOTE: Within group "location=A" we uniformly sampled across "year"
+ however there were insufficient samples for "year=2002" resulting in
+ in the output distribution not matching the input distribution
```

Reply:
This odd behavior already exists for current "uniform" sampling (the behavior when `--group-by-weights` is removed from this example). The weighting is slightly different in that it will weight evenly across all permutations of `year` and `location` values that have at least one sequence in the metadata. In other words, it would be the equivalent of specifying this weights file:

```sh
cat >weights.tsv <<~~
year	location	weight
2000	A	1
2001	A	1
2002	A	1
2000	B	1
2001	B	1
2002	B	1
~~
```

which would, before adding the `2002A_1	2002	A` row, produce this `--output-group-by-sizes` file:

```
year	location	weight	_augur_filter_target_size_int
2000	A			1		20
2001	A			1		20
2000	B			1		20
2001	B			1		20
2002	B			1		20
```

and after adding that row produce this:

```
year	location	weight	_augur_filter_target_size_int
2000	A			1		17
2001	A			1		16
2002	A			1		16
2000	B			1		16
2001	B			1		17
2002	B			1		17
```

This gets back to the point in https://github.com/nextstrain/augur/pull/1454#discussion_r1607249501 about adding number of outputted sequences to this output file. Example:

```
year	location	weight	_augur_filter_target_size_int	sequences_that_passed_filter
2000	A			1		0								20
2000	B			1		0								20
2001	A			1		0								20
2001	B			1		0								20
2002	B			1		0								20
```

```
year	location	weight	_augur_filter_target_size_int	sequences_that_passed_filter
2000	A			1		17								17
2001	A			1		16								16
2002	A			1		16								1
2000	B			1		16								16
2001	B			1		17								17
2002	B			1		17								17
```

The missing piece there (and also what's needed to produce the additional output in your suggestion) is to update this function to be aware of the number of rows per group available in the metadata. It should be doable, I will take a look. Once that information is availalbe we can also consider the in-filling approach you've described.

Reply:
I'm not entirely convinced that in-filling is the best approach, it was just what I came up with upon review, so perhaps don't invest time in that implementation until others have expressed their thoughts. But having warning/log messages printed in these cases will be helpful (assuming one reads the warning messages).

Reply:
I made a similar note in the description of https://github.com/nextstrain/ncov/pull/1106:

> Caveat: weighted target sizes are calculated without taking into account the actual number of sequences available per group. This means under-sampled countries will still be under-sampled even if it has a high weight, and results in significantly fewer total sequences than requested by `--subsample-max-sequences`. This is already the case with the uniform sampling approach prior to this PR, but it may be more noticeable now with large countries that are under-sampled.

Reply:
Added better outputs in 3c64691f5e853b844d69c4cfbf7fb765bdb58449. Will test in https://github.com/nextstrain/ncov/pull/1106

Reply:
This is what it looks like, from [run](https://github.com/nextstrain/ncov/actions/runs/10294116138/job/28491628027#step:5:793):

```
WARNING: Targeted 19 sequences for group ["country='China'", 'year=2017', "month='2017-8'"] but only 1 is available.
WARNING: Targeted 19 sequences for group ["country='China'", 'year=2018', "month='2018-8'"] but only 1 is available.
WARNING: Targeted 18 sequences for group ["country='China'", 'year=2019', "month='2019-10'"] but only 1 is available.
WARNING: Targeted 20 sequences for group ["country='India'", 'year=2020', "month='2020-1'"] but only 2 are available.
WARNING: Targeted 2 sequences for group ["country='Bangladesh'", 'year=2020', "month='2020-2'"] but only 1 is available.
```

It was easy to implement and is more verbose than your initial suggestion, but lacks info on weighted vs. unweighted columns. That could be added:

```
WARNING: Undersampling occurred for the following groups within ('country'='China'):
         - Targeted 19 sequences for ('year'='2017', 'month='2017-8') but only 1 is available
         - Targeted 19 sequences for ('year'='2018', 'month='2018-8') but only 1 is available
         - Targeted 18 sequences for ('year'='2019', 'month='2019-10') but only 1 is available
WARNING: Undersampling occurred for the following groups within ('country'='India'):
         - Targeted 18 sequences for ('year'='2020', 'month='2020-1') but only 2 are available
WARNING: Undersampling occurred for the following groups within ('country'='Bangladesh'):
         - Targeted 2 sequences for ('year'='2020', 'month='2020-2') but only 1 is available
```

The output would be simpler for the [other case](https://github.com/nextstrain/augur/pull/1454#discussion_r1704761006) (weighted columns only):

```
WARNING: Undersampling occurred for ('location'='A'): targeted 100 sequences but only 50 are available.
```

Reply:
I agree that this is a tricky situation. In the original formulation of only `--sequences-per-group` the idea was to say specify `--sequences-per-group 10` and `--group-by country` would target 10 sequences per country and randomly sample these sequences for each country group. In the original formulation, we wouldn't top-up other countries. I think this is a semantic complication with adding the convenience parameter of `--subsample-max-sequences`. I'd think of `--subsample-max-sequences` and solely specifying `--sequences-per-group`.

All this to say, that I believe the current behavior of 
```
WARNING: Targeted 19 sequences for group ["country='China'", 'year=2017', "month='2017-8'"] but only 1 is available.
```
is exactly what we want. It's okay if there's holes in the resulting sampling if the underlying data doesn't support filling all groups. It's a bit hard to predict what feels most natural to everyone, but this behavior seems like the least surprising behavior to me.

Reply:
Great, I'll leave the output as-is. We can tweak as we start using this in workflows.